### PR TITLE
Shadow support for color contrast

### PR DESCRIPTION
**Pull request by marcysutton**
This PR adds color contrast support for pages with nested Shadow DOM. 

One TODO I still have is to support multiline elements inside of shadow boundaries (inline elements with line breaks, originally added in https://github.com/dequelabs/axe-core/pull/610). The actual background color for those is wrong in this PR because the tree walking across multiple roots doesn't really factor in client rects that partially overlap. I've struggled with it for the past week and decided to punt on it.

Part of the complication is that a shadow host is always returned from `document.elementsFromPoint` regardless of whether its content overlaps (i.e. absolute positioning, etc.). For block elements that don't visually overlap their host element, we have to filter out the host element from the stack or it ends up in Needs Review for `elmPartiallyObscured`. But then we're in a conundrum when multiline elements only partially overlap: our utilities for overlap checking only look at the boundingClientRect, and it would take some refactoring to check individual rects against a shadow host. 

I sorta wonder if the `contentOverlapping` method in `axe.commons.color.getBackgroundColor` should just go away....it seems like we're checking the same thing more than once, and it's way more complicated now (checking overlap of individual rects against multiple bgNodes, across multiple roots). Some of the work has been moved to the stack gathering steps, where we did more evaluation after the fact before. There's probably some redundancy that could get cleaned up as a next step.

Closes https://github.com/dequelabs/axe-core/issues/687

### CODE REVIEW

File: test/commons/color/get-background-color.js [hunk 1]
```diff
@@ -626,4 +648,146 @@ describe('color.getBackgroundColor', function () {
...
+		var actual = axe.commons.color.getBackgroundColor(target, [], false);
+
+		assert.closeTo(actual.red, 0, 0);
+		assert.closeTo(actual.green, 0, 0);
+		assert.closeTo(actual.blue, 0, 0);
+		assert.closeTo(actual.alpha, 1, 0);
+	});
+
+	(shadowSupported ? it : xit)
+	('should count an implicit label as a background element inside shadow dom', function () {
```
Comment (WilcoFiers) [inline test/commons/color/get-background-color.js]: You should use a different colour so you can distinguish it from the background.

File: test/commons/color/get-background-color.js [hunk 2]
```diff
@@ -626,4 +648,146 @@ describe('color.getBackgroundColor', function () {
...
+		assert.closeTo(actual.blue, 255, 0);
+		assert.closeTo(actual.alpha, 1, 0);
+	});
+
+	(shadowSupported ? it : xit)
+	('finds a color for absolutely positioned content when background is in shadow dom', function () {
+		fixture.innerHTML = '<div id="elm1" style="width:10em; height:0; position:absolute;"></div>' +
+			'<div id="elm2" style="color:green; position:absolute;">Text</div>';
+
+	  var elm1 = document.querySelector('#elm1');
```
Comment (WilcoFiers) [inline test/commons/color/get-background-color.js]: Mixing tabs and spaces.

File: test/commons/color/get-background-color.js [hunk 3]
```diff
@@ -626,4 +648,146 @@ describe('color.getBackgroundColor', function () {
...
+	(shadowSupported ? it : xit)
+	('finds colors for absolutely positioned elements across shadow boundaries', function () {
+		fixture.innerHTML = '<div id="container" style="background-color:black; height:20px; position:relative;"></div>';
+		var container = fixture.querySelector('#container');
+		var shadow = container.attachShadow({ mode: 'open' });
+		shadow.innerHTML = '<div id="shadowTarget" style="color:#333; height:20px; position:absolute; top:20px;">Text</div>';
+
+		var target = shadow.querySelector('#shadowTarget');
+		var actual = axe.commons.color.getBackgroundColor(target, []);
+		assert.closeTo(actual.red, 255, 0);
```
Comment (WilcoFiers) [inline test/commons/color/get-background-color.js]: Why did you use `assert.closeTo()` here? If your margin is 0 you can use `assert.equal`, right? Same question for some of the other tests below.

File: test/commons/color/get-background-color.js [hunk 4]
```diff
@@ -626,4 +648,146 @@ describe('color.getBackgroundColor', function () {
...
+		fixture.innerHTML = '<div id="container" style="background-color:black; height:20px;"></div>';
+		var container = fixture.querySelector('#container');
+		var shadow = container.attachShadow({ mode: 'open' });
+		shadow.innerHTML = '<div id="shadowTarget" style="color:#333;">Text<br>More text</div>';
+
+		var target = shadow.querySelector('#shadowTarget');
+		var actual = axe.commons.color.getBackgroundColor(target, []);
+		assert.isNull(actual);
+	});
 });
```
Comment (WilcoFiers) [inline test/commons/color/get-background-color.js]: We should test slotted text with a background inside a shadow tree.

File: test/commons/color/get-background-color.js (hunk 3 above)
Comment (marcysutton) [reply test/commons/color/get-background-color.js]: I'd copied it from the existing tests, I can change it!

File: test/commons/color/get-background-color.js (hunk 2 above)
Comment (marcysutton) [reply test/commons/color/get-background-color.js]: Thanks Sublime 😡

File: lib/commons/color/get-background-color.js [hunk 5]
```diff
@@ -204,7 +205,33 @@ color.getCoords = function(rect) {
...
- * Get elements from point for block and inline elements, excluding line breaks
+ * Get elements from point across shadow dom boundaries
+ * @method getElementsFromPoint
+ * @memberof axe.commons.color
+ * @instance
+ * @param {Object} nodeCoords X and Y coordinates of point
+ * @param {Object} [root] Shadow root or document root
+ * @return {Array}
+ */
+color.getElementsFromPoint = function(nodeCoords, root = document) {
```
[the remark this answers is not in the document]
Comment (marcysutton) [reply lib/commons/color/get-background-color.js]: Done.

[the remark this answers is not in the document]
Comment (marcysutton) [reply lib/commons/color/get-background-color.js]: However, in doing so, I added some tests that are apparently pretty brittle on Circle. they all pass locally. I don't think we should sink a bunch of time into this if the API is tested in a bunch of other places (getBackgroundColor, contrast check tests).

**Comment by marcysutton** [comment]
I can't get that chrome:mobile test to fail locally even in the Responsive Design viewer at that viewport, so it's a Circle-only error. I'm also unable to debug Circle with Xquartz any longer...something is borked in that process for me. When I try to run Chrome or `xclock` I get `"Can't open display: localhost:10.0"` with very little permissions to fix it. Searching didn't bring up any promising debugging tips. https://github.com/dequelabs/axe-core/blob/develop/CONTRIBUTING.md#debugging-tests-that-only-fail-on-circleci

**Pull request merged**